### PR TITLE
Duration and duration-mode options 

### PR DESCRIPTION
**Pull request by alanconway**
only before/after-send/receive are implemented
the -tx-action variants are ignored

### CODE REVIEW

**Comment by rkubis** [comment]
I tried this and code works fine, but all messages are still sent after all waitings, not in the background of waiting.

The proposed behaviour needs `schedule` or something like this?

**Comment by alanconway** [comment]
Yes you are right, the sleeps are blocking the IO. I  will fix with schedule today

**Comment by alanconway** [comment]
Fixed to use schedule so send is timed correctly. Added a TimedServer test to verify messages are timed correctly as received by the server.

**Comment by rkubis** [comment]
Closes #18

File: lib/handlers/receiver_handler.rb [hunk 1]
```diff
@@ -93,6 +96,8 @@ def initialize(
       @sent = 0
       # Counter of accepted messages
       @accepted = 0
+      # Duration
+      @duration = Duration.new(duration, count, duration_mode);
```
Comment (rkubis) [inline lib/handlers/receiver_handler.rb:100]: I will remove semicolon.

File: lib/handlers/sender_handler.rb [hunk 2]
```diff
@@ -130,6 +133,10 @@ def initialize(
       @sent = 0
       # Number of accepted messages
       @accepted = 0
+      # Duration
+      @duration = Duration.new(duration, count, duration_mode);
```
Comment (rkubis) [inline lib/handlers/sender_handler.rb:137]: I will remove semicolon.

File: lib/handlers/sender_handler.rb [hunk 3]
```diff
@@ -158,62 +165,77 @@ def on_container_start(container)
         })
     end
 
+    def delay
+      before = @duration.delay("before-send")
+      after = @duration.delay("after-send") if @sent > 0 # No after-delay on first send
+      [before, after].compact.sum
```
Comment (rkubis) [inline lib/handlers/sender_handler.rb:171]: Array#sum is not supported in Ruby 2.0.0, I will change this to Array#inject(:+).

File: lib/handlers/sender_handler.rb [hunk 4]
```diff
@@ -158,62 +165,77 @@ def on_container_start(container)
...
-        msg = Qpid::Proton::Message.new
-        msg.address = @broker.amqp_address if @anonymous
-        # Set message properties
-        if @msg_properties
-          @msg_properties.each { |k, v| msg[k] = v }
+      if @duration.zero?        # Send immediately
+        send(sender) while (sender.credit > 0) && (@sent < @count)
+      elsif (sender.credit > 0) && (@sent < @count) && !@scheduled # Schedule to send after delay
+        @scheduled = true
+        c = sender.connection.container;
```
Comment (rkubis) [inline lib/handlers/sender_handler.rb:182]: I will remove semicolon.

**Review by rkubis — APPROVED** [review]
I will modify some details, otherwise looks good.

**Comment by rkubis** [comment]
Thanks a million, Alan!

**Pull request merged**